truffle/src/main/java/org/jruby/truffle/runtime/rope/Rope.java

```
@@ -93,4 +106,34 @@ public int realSize() {
     public int getRealSize() {
         return realSize();
     }
+
+    @Override
+    public int hashCode() {
+        if (hashCode == 0) {
+            hashCode = RopeOperations.hashForRange(this, 1, 0, byteLength);
+        }
+
+        return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o instanceof Rope) {
```

I like fail fast on error-like values, but it's just a preference.